### PR TITLE
pfblockerng: case-fold the TOP1M TLD inclusion match; anchor the csv hostname guard

### LINKED ISSUE #920: pfblockerng: TOP1M csv guard — case-fold the TLD inclusion match; consider the /D modifier
Two pre-existing (not introduced by #913) items surfaced by the adversarial review of PR #913, tracked here per the review-triage rule.

1. **Case-sensitivity mismatch in the TLD inclusion match.** In `pfblockerng_top1m()` (`src/usr/local/pkg/pfblockerng/pfblockerng.inc`, just below the csv-mode guard), `$tld = substr($domain, strrpos($domain, '.') + 1)` is checked via `isset($pfb_include[$tld])` with no case-folding, while the include keys come from the GUI's fixed lowercase list. A feed row with an uppercase/mixed-case TLD (`EXAMPLE.XN--P1AI`, `EXAMPLE.COM`) passes the (case-insensitive) hostname regex and counts toward `$linecnt`, but silently fails the whitelist match. Real-world TOP1M CSVs are effectively always lowercase, so this is theoretical today — fix is a `strtolower()` at the compare.

2. **No `D` modifier on the hostname guard regex.** PHP PCRE's `$` without `/D` matches immediately before a single trailing `\n`; only plausibly reachable for a single-column mode where the domain field could retain a trailing newline. Same behaviour as the pre-#913 regex.

Both equally affected the code before PR #913; neither blocks it.

---
🤖 Generated by [Claude Code](https://claude.com/claude-code), posted via @andrebrait's account on their behalf.

### LINKED ISSUE #950: dev-tooling: composer install fails in Anthropic cloud sessions (api.github.com egress-scoped) — automate the fallback
In Claude Code managed cloud sessions, `composer install` fails out of the box and costs significant session time to work around by hand. Diagnosed 2026-07-07 (session working issue #920):

## Root cause (probed, not guessed)

1. The session's GitHub egress is **repo-scoped**: `api.github.com` (and `codeload.github.com`) return `403` `{"message":"GitHub access to this repository is not enabled for this session. Use add_repo to request access."}` for every third-party repo. Every composer **dist** (zipball) download therefore fails with `Could not authenticate against github.com`.
2. Plain `git` clone/fetch of arbitrary public GitHub repos **works** (only the REST/codeload endpoints are scoped), so `composer install --prefer-source` succeeds for every package **except**:
3. `phpstan/phpstan`, which is **dist-only** in `composer.lock` (no `source` entry — it is the phar distribution repo), and whose full mirror clone (`git clone --mirror`, what composer does for source installs) **times out at 300 s** (phar blobs across the whole history).

## Working recipe (validated in-session)

1. Temporarily strip `phpstan/phpstan` from `composer.json` (require-dev) + `composer.lock` (composer hard-errors on a lock/json mismatch, so both).
2. `COMPOSER_ALLOW_SUPERUSER=1 composer install --no-interaction --prefer-source` — installs everything else via git.
3. Restore both files (`git checkout composer.json composer.lock`).
4. Shallow-fetch phpstan at the lock's pinned reference (fast — single commit, no history):

   ```sh
   git init tmp && git -C tmp fetch --depth 1 https://github.com/phpstan/phpstan.git <lock dist.reference>
   git -C tmp checkout FETCH_HEAD
   ```

5. Copy into `vendor/phpstan/phpstan` (drop `.git`), symlink `vendor/bin/phpstan -> ../phpstan/phpstan/phpstan`.

All gates then run normally (`phpunit` 2376 tests, `phpstan 2.2.1`, `phpcs`).

## Ask

Automate this as a dev-tooling script (e.g. `scripts/composer-cloud-install.sh`): detect the plain-`composer install` failure (or detect the scoped-egress 403 up front), then apply the fallback above, deriving the phpstan ref from `composer.lock` instead of hardcoding it. Optionally wire it into the repo's Claude cloud SessionStart path so sessions stop rediscovering this.

Known environment caveat for test runs in these sessions (unrelated to composer, recorded here so it's greppable): the container runs as **root**, so `Top1mPreserveOnEmptyFeedTest::testReadOnlyDbdirCausesRenameFailurePreservesPriorWhitelistAndWarns` fails on a clean checkout (root bypasses the chmod-0555 read-only dbdir). CI (non-root) is unaffected.

### PR DESCRIPTION
**Pull request by andrebrait**
Fixes #920. Fixes #950.

## TOP1M csv guard (#920) — both items confirmed by in-session probes

1. **Case-fold the TLD inclusion match.** `pfblockerng_top1m()` compared `isset($pfb_include[$tld])` case-sensitively while the include keys are always lowercase (the GUI's fixed `$options_alexa_inclusion` list). An uppercase/mixed-case feed domain (`EXAMPLE.COM`, `example.XN--P1AI`) passed every guard and counted toward `$linecnt` but silently failed the whitelist match. Fix: `strtolower()` the extracted domain once, so the TLD compare, the `www.` strip, and the written whitelist entries are all canonical in one spot (downstream `pfb_unbound.py` already lowercases on ingest; real-world feeds are lowercase, so current inputs are byte-identical).
2. **`D` modifier on the csv-mode hostname guard.** Probed: `str_getcsv()` strips a *bare* trailing newline in every mode, but a **quoted unterminated** field (a truncated/corrupt feed line) retains it — and without `/D`, `$` matches before that newline, so the garbage row passed the guard, incremented `$linecnt`, and could flip a garbage-only feed from dead (preserve prior whitelist) to valid (swap in an empty build) — the same failure class as the #886/#892 guards. This narrows the issue's original "only plausibly reachable for a single-column mode" guess: the reachable path is a quoted unterminated field in **any** `csv`-parse provider.

New `tests/php/Top1mTldCaseFoldTest.php` (9 tests) pins both fixes across every parse mode × domain column from the provider registry (`rank_domain` col 1; `csv` cols 0/1/2 — Cloudflare/OpenPageRank/Majestic shapes), including the negative guard (fold must not over-match), the lowercase control (no regression), and both quoted-unterminated dead-feed-preserve transitions with before-state assertions. Red→green executed: on the pre-change source the 7 fix-dependent tests fail, the 2 fix-independent guards pass; all 9 green post-fix. Full PHP gates green (`php -l`, PHPUnit 2385 tests, PHPStan, PHPCS).

## Composer bootstrap for managed cloud sessions (#950)

`scripts/composer-cloud-install.sh` automates the validated fallback for egress-restricted cloud sessions where `api.github.com`/`codeload.github.com` are repo-scoped (403 for third-party dists): plain `composer install` first; on failure, strip dist-only `phpstan/phpstan` from `composer.json`+`composer.lock`, install the rest `--prefer-source` over git, restore both files, then shallow-fetch phpstan at the lock's pinned reference into `vendor/`. Offline-testable helpers covered by `tests/shell/composer_cloud_install_spec.sh` (shellspec, 4 examples; full shell suite 387/0); `sh -n` + shellcheck clean. One-line pointer added to CLAUDE.md "Running tests".

Bundled on one branch because this managed session's push policy pins all work to a single minted branch; the two changes are otherwise independent commits.

---
_Generated by [Claude Code](https://claude.ai/code/session_01CWzcfy2513ohNTkSVHm8WF)_

## Summary by CodeRabbit

* **New Features**
  * Added a fallback installation path for Composer in managed-cloud environments when the standard install hits a 403 error.

* **Bug Fixes**
  * Improved domain validation so mixed-case entries are handled consistently and malformed CSV rows are rejected more reliably.
  * Preserves the existing TOP1M whitelist when new feed data contains invalid or unterminated fields.

* **Documentation**
  * Updated test/setup instructions to include the new Composer fallback command.

* **Tests**
  * Added coverage for case-folding, CSV parsing edge cases, and the Composer fallback workflow.

### CODE REVIEW

File: scripts/composer-cloud-install.sh [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+	fi
+
+	echo "composer-cloud-install: plain install failed -- applying the cloud egress fallback (issue #950)" >&2
+
+	# The fallback mutates composer.json/lock and restores them via git checkout;
+	# refuse to run over uncommitted user edits it would clobber.
+	if ! git diff --quiet -- composer.json composer.lock; then
+		echo "composer-cloud-install: composer.json/composer.lock have uncommitted changes; commit or stash them first" >&2
+		exit 1
+	fi
```
[the remark this answers is not in the document]
Comment (andrebrait) [reply scripts/composer-cloud-install.sh]: Applied in 675ec94 (rebased). The guard now diffs against HEAD (`git diff --quiet HEAD -- composer.json composer.lock`), so staged edits are refused too; extracted into `refuse_if_composer_files_dirty()` with shellspec coverage for the clean/unstaged/staged cases — the staged case was red against the old index-only guard (executed mutation check). One precision on the premise: `git checkout -- <paths>` restores from the *index*, so staged edits would not actually have been discarded — but the guard's intent is "no uncommitted edits at all", and the staged case did wrongly pass it, hence the fix.

---
🤖 Generated by [Claude Code](https://claude.com/claude-code), posted via @andrebrait's account on their behalf.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: scripts/composer-cloud-install.sh [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+#   1. Plain `composer install`; on success, done — no fallback needed.
+#   2. Strip phpstan/phpstan from composer.json + composer.lock (composer
+#      hard-errors on a json/lock mismatch, so both), then
+#      `composer install --prefer-source` (git transport), then restore both files.
+#   3. Shallow-fetch phpstan at the lock's pinned dist reference (single commit,
+#      fast), place it under vendor/phpstan/phpstan, symlink vendor/bin/phpstan.
+#
+# ponytail: phpstan/phpstan is the only dist-only package in today's lock;
+# generalize the strip/fetch to a package list if another one ever appears.
+set -eu
```
[the remark this answers is not in the document]
Comment (andrebrait) [reply scripts/composer-cloud-install.sh]: Applied in 675ec94 (rebased): top-level `set -eu` removed; strict mode is now the first line of `main()`, which only runs behind the existing `case "${0##*/}"` execution guard, so sourcing the file (as the shellspec suite does) no longer mutates the caller's shell options.

---
🤖 Generated by [Claude Code](https://claude.com/claude-code), posted via @andrebrait's account on their behalf.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: scripts/composer-cloud-install.sh [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+		$l = json_decode(file_get_contents($argv[2]), true);
+		foreach (["packages", "packages-dev"] as $k) {
+			$l[$k] = array_values(array_filter($l[$k] ?? [], fn ($p) => $p["name"] !== "phpstan/phpstan"));
+		}
+		file_put_contents($argv[2], json_encode($l, JSON_PRETTY_PRINT | JSON_UNESCAPED_SLASHES) . "\n");
+	' "$1" "$2"
+}
+
+main() {
+	cd "$REPO_ROOT"
```
[the remark this answers is not in the document]
Comment (andrebrait) [reply scripts/composer-cloud-install.sh:71]: Applied in 675ec94 (rebased) together with the previous comment — `set -eu` is inside `main()`, so direct execution still fails fast (verified with a POSIX-sh probe that a helper's `return 1` under `set -e` aborts with exit 1, preserving the old `exit 1` behaviour).

---
🤖 Generated by [Claude Code](https://claude.com/claude-code), posted via @andrebrait's account on their behalf.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: scripts/composer-cloud-install.sh [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+		exit 1
+	fi
+
+	trap 'git -C "$REPO_ROOT" checkout -- composer.json composer.lock' EXIT
+	strip_phpstan composer.json composer.lock
+	composer install --no-interaction --prefer-source
+	git checkout -- composer.json composer.lock
+	trap - EXIT
+
+	staging="$(mktemp -d)"
```
[the remark this answers is not in the document]
Comment (andrebrait) [reply scripts/composer-cloud-install.sh]: Applied in 675ec94 (rebased): `mktemp -d "${TMPDIR:-/tmp}/composercloud.XXXXXX"` — matches the template convention already used in `scripts/local-smoke.sh` and `pfblockerng.sh`.

---
🤖 Generated by [Claude Code](https://claude.com/claude-code), posted via @andrebrait's account on their behalf.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by andrebrait** [comment]
## Review audit trail

**Reviews received.** Claude Sonnet 5 adversarial review ran as the `review-fanout` workflow (3 independent lenses → dedup → execution-grounded verification of every finding; 9 agents, effort `xhigh`) — chosen over a single-agent pass because the PR exceeds 300 changed lines and touches parsing/guard behaviour in `src/`. GitHub Copilot reviewed (4 inline comments). CodeRabbit was **skipped under the 5-minute rule**: its rate-limit notice stated "Next review available in: 40 minutes". Snyk: no terminal failure verdict on the head SHA (see checks).

**Findings ledger** (fanout = review-fanout confirmed findings; C = Copilot):

1. fanout/blocking — csv case-fold proven against the real registered descriptor only for Majestic (col 2); Cloudflare (col 0) tested via a synthetic `header=FALSE` stand-in no real provider matches; OpenPageRank (col 1) had no case-fold assertion → **fixed@675ec94**: rows 6–7 now use the real Cloudflare descriptor (header line in fixture), new `testCsvDomainCol1OpenPageRankShapeUppercaseDomainMatchesAndFolds` covers col 1. Red→green re-executed: 8/10 red on the pre-fix source, 10/10 green after.
2. fanout/blocking + C1 — composer.json/lock dirty guard untested and index-only (staged edits passed it) → **fixed@675ec94**: extracted `refuse_if_composer_files_dirty()`, diffs against HEAD, 3 shellspec examples (clean/unstaged/staged; staged case executed red against the old guard).
3. fanout/blocking — script committed 100644 while documented for direct invocation → **fixed@675ec94** (mode 100755).
4. C2+C3 — top-level `set -eu` mutates the sourcing shell → **fixed@675ec94** (strict mode inside `main()`).
5. C4 — bare `mktemp -d` not portable → **fixed@675ec94** (template, house convention).
6. fanout/outside-diff — pre-existing: `rank_domain` parse has no hostname-shape guard; a dotless domain field yields a garbage TLD via `strrpos()` false → **deferred: #954** (unchanged by this diff, out of #920's scope).
7. fanout/nitpick — rows 6–7 docblock named Cloudflare while using a synthetic descriptor → **fixed@675ec94** (subsumed by item 1).

**Focused re-review of the fix delta** (675ec94, new unreviewed code): independent Sonnet 5 pass at `xhigh` — verdict CLEAN, all findings confirmed closed via executed mutation checks (dropping `strtolower()` fails the three per-column fold tests; reverting the HEAD-diff guard fails the staged-edits example).

**Gates on the final head**: PHPUnit 2386 tests — sole failure is the documented root-container baseline (`testReadOnlyDbdirCausesRenameFailurePreservesPriorWhitelistAndWarns`, root bypasses chmod-0555; CI runs non-root); PHPStan clean; PHPCS clean; `php -l` clean; `sh -n` + shellcheck clean; shellspec 436 examples / 0 failures.

---
🤖 Generated by [Claude Code](https://claude.com/claude-code), posted via @andrebrait's account on their behalf.

---
_Generated by [Claude Code](https://claude.ai/code)_
